framework/src/main/java/org/checkerframework/common/accumulation/AccumulationTransfer.java

```diff
@@ -0,0 +1,152 @@
+package org.checkerframework.common.accumulation;
+
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.tree.Tree.Kind;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import javax.lang.model.element.AnnotationMirror;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.checkerframework.common.value.ValueCheckerUtils;
+import org.checkerframework.dataflow.analysis.FlowExpressions;
+import org.checkerframework.dataflow.analysis.FlowExpressions.Receiver;
+import org.checkerframework.dataflow.analysis.TransferResult;
+import org.checkerframework.dataflow.cfg.node.MethodInvocationNode;
+import org.checkerframework.dataflow.cfg.node.Node;
+import org.checkerframework.framework.flow.CFAnalysis;
+import org.checkerframework.framework.flow.CFStore;
+import org.checkerframework.framework.flow.CFTransfer;
+import org.checkerframework.framework.flow.CFValue;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+
+/**
+ * The default transfer function for an accumulation checker.
+ *
+ * <p>Subclasses should call the {@link #accumulate(Node, TransferResult, String...)} accumulate}
+ * method to accumulate a string at a particular program point.
+ */
+public class AccumulationTransfer extends CFTransfer {
+
+    /** The type factory. */
+    protected final AccumulationAnnotatedTypeFactory typeFactory;
+
+    /**
+     * Build a new AccumulationTransfer for the given analysis.
+     *
+     * @param analysis the analysis
+     */
+    public AccumulationTransfer(CFAnalysis analysis) {
+        super(analysis);
+        typeFactory = (AccumulationAnnotatedTypeFactory) analysis.getTypeFactory();
+    }
+
+    /**
+     * Updates the estimate of how many things {@code node} has accumulated.
+     *
+     * <p>If the node is an invocation of a method that returns its receiver, then its receiver's
+     * type will also be updated. This chain will continue as long as each receiver is itself a
+     * receiver-returning method invocation.
+     *
+     * <p>For example, suppose the argument is the expression {@code a.b().c()}, the new value is
+     * "foo", and b and c return their receiver (and are deterministic). Then all of the expressions
+     * {@code a}, {@code a.b()}, and {@code a.b().c()} would have their estimates updated to include
+     * "foo".
+     *
+     * <p>If any method in the chain is non-deterministic, its estimate will not be updated (but the
+     * types of any deterministic methods, and the ultimate receiver, will be updated).
```

Comment:
"any deterministic methods": they might follow it, as in `a.b().nondet().c().d()` where b, c, and d are deterministic.  Do the values go through `nondet()` and affect all of b(), c(), and d()?

Reply:
yes, as long as `nondet` returns `@This`. I think this is just an imprecision caused by the dataflow framework rather than something fundamental, since the RR information should make the fact that the method isn't known to be deterministic irrelevant.